walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +40,263 @@ public int compare(Event a, Event b) {
-   * or more time slots exists so that both mandatory and optional attendees can attend, it returns
-   * those time slots. Otherwise, it returns the time slots that fit just the mandatory attendees.
-   *
-   * @param eventsCollection the events we know about
-   * @param request information about the meeting, including attendees, optional attendees, and how
-   *     long it needs to be
-   */
-  public Collection<TimeRange> query(Collection<Event> eventsCollection, MeetingRequest request) {
-    Collection<TimeRange> withOptionalAttendees = getMeetingTimes(eventsCollection, request, true);
-
-    // Special case: if no mandatory attendees and optional attendees' schedules cannot fit in a
-    // meeting, no meeting times are possible.
-    return !withOptionalAttendees.isEmpty() || request.getAttendees().isEmpty()
-        ? withOptionalAttendees
-        : getMeetingTimes(eventsCollection, request, false);
-  }
+  // All logic needed to find optimal meeting times with one instance of meeting information and
+  // events to be considered.
+  private static class SingleMeetingResolver {
+    // all events to be considered
+    private final Collection<Event> events;
+
+    // information about the meeting, including attendees, optional attendees, and how long it needs
+    // to be
+    private final MeetingRequest request;
 
-  private Collection<TimeRange> getMeetingTimes(
-      Collection<Event> eventsCollection,
-      MeetingRequest request,
-      boolean includeOptionalAttendees) {
-    HashSet<String> attendees = new HashSet<String>(request.getAttendees());
-    if (includeOptionalAttendees) {
-      attendees.addAll(request.getOptionalAttendees());
+    //
+    private ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
+
+    SingleMeetingResolver(Collection<Event> events, MeetingRequest request) {
+      this.events = events;
+      this.request = request;
     }
-    ArrayList<Event> events = getRelevantEvents(attendees, new ArrayList<Event>(eventsCollection));
-    List<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
 
-    // Need to check this so we don't access out of bounds when we add first gap.
-    if (events.isEmpty()) {
-      addIfLongEnough(
-          TimeRange.fromStartEnd(0, END_OF_DAY, true), possibleMeetingTimes, request.getDuration());
-      return possibleMeetingTimes;
+    /**
+     * Returns a list of time periods in which the meeting, specified by request, could happen. If
+     * no time exists for all optional and mandatory attendees, find the time slot(s) that allow
+     * mandatory attendees and the greatest possible number of optional attendees to attend.
+     */
+    Collection<TimeRange> resolveBestTime() {
+      ArrayList<TimeRange> mandatoryAttendeesMeetingTimes =
+          getMandatoryAttendeesMeetingTimes(
+              request.getDuration(), new HashSet<String>(request.getAttendees()));
+      Collection<TimeRange> optimalMeetingTimes =
```

Comment:
getOptimalMeetingTimes is returning the member variable `optimalMeetingTimes`, so you don't need to return or assign it here.